spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -308,6 +332,29 @@ private ReadTask(CombinedScanTask task, String tableSchemaString, String expecte
         encryptionManager.value(), caseSensitive);
     }
 
+    @Override
+    public String[] preferredLocations() {
+      if (!localityPreferred) {
+        return new String[0];
+      }
+
+      Configuration conf = SparkSession.active().sparkContext().hadoopConfiguration();
+      Set<String> locations = Sets.newHashSet();
+      for (FileScanTask f : task.files()) {
+        Path path = new Path(f.file().path().toString());
+        try {
+          FileSystem fs = path.getFileSystem(conf);
+          for (BlockLocation b : fs.getFileBlockLocations(path, f.start(), f.length())) {
+            locations.addAll(Arrays.asList(b.getHosts()));
+          }
+        } catch (IOException ioe) {
+          LOG.warn("Failed to get block locations for path {}", path, ioe);
```

Comment:
Since `localityPreferred` flag is `true`.  We should throw a `org.apache.iceberg.exceptions.RuntimeIOException` here. 
Also we can use `Util.getFS` to get the FileSystem class

Reply:
I think I agree with the existing warn behavior. If we can't get locality info from the FS it shouldn't fail the job. It is a preference, not a requirement right?